### ISSUE TITLE
multi-node support

### ISSUE THREAD

**Issue by astooke** [astooke/synkhronos#12]
Starting a new issue in reference to question: (https://github.com/astooke/Synkhronos/issues/11#issuecomment-326628646)

I have not experimented with running Synkhronos multi-node.  Currently it's only built for single-node.  To run multi-node would require another layer to coordinate and communicate among nodes.  Certainly sounds possible, with a separate instance of the current Synkhronos running on each node.  I haven't put a lot of thought into this yet, because my current research is well-suited to running single-node.  

Apparently the new version of NCCL, version 2, supports inter-node communication.  I have not tried it yet (Synkhronos is currently built on version 1).  Synkhronos uses NCCL through libgpuarray and pygpu...I'm not sure what the compatibility status is through that chain.

Note that a key to scaling well to 256 GPUs in the [large minibatch ResNet paper](https://arxiv.org/pdf/1706.02677.pdf) is to start communicating on gradients as they are computed layer-by-layer, simultaneously with performing the rest of the backpropagation.  

I'd be curious to hear if you try anything!

Have you tried any other packages / libraries for running multi-GPU?  e.g. TensorFlow, PyTorch, Chainer?  And how does using them compare to Synkhronos?

**Comment by nouiz** [comment]
for libgpuarray, the new version 0.7 (and 0.7.1 to be released today to fix
some compilation crash with cuda8) need NCCL2. So someone can start to
explore how to use it for multi-node computation.

Note, Theano master don't support that version of pygpu yet. We need to
merge a PR first. We are waiting for 0.7.1 to be out for that.

**Comment by Nqabz** [comment]
I recently tried NCCLv2 but  discovered that I pygpu was not updated to match the API changes.
NVIDIA then recently fixed some GPUArray issues for compatibility with NCCLv2 and they are waiting to commit sometime soon. Not sure if that will be internally or open source? Will post back once I have more information.

**Comment by nouiz** [comment]
Note,

If you install yourself libgpuarray and pygpu from the master of
libgpuarray it will work.

But you will need to install this branch of Theano:

https://github.com/Theano/Theano/pull/6317

and rebase it too. This should be merged next week. (I already wrote that,
but this should be real this time:)

So maybe wait next week, it would be simpler.

**Comment by Nqabz** [comment]
@nouiz Sounds great. I will await your merge next week.

Thanks!
